### PR TITLE
u_set replaced by indexed_uint_set

### PR DESCRIPTION
**Pull request by levnach**

### CODE REVIEW

File: src/math/lp/lar_solver.cpp [hunk 1]
```diff
@@ -1739,8 +1731,10 @@ namespace lp {
         TRACE("lar_solver_feas", tout << "j = " << j << " became " << (this->column_is_feasible(j)?"feas":"non-feas") << ", and " << (this->column_is_bounded(j)? "bounded":"non-bounded") << std::endl;);    
     }
     void lar_solver::insert_to_columns_with_changed_bounds(unsigned j) {
-        m_columns_with_changed_bounds.insert(j);
-        TRACE("lar_solver", tout << "column " << j << (column_is_feasible(j) ? " feas" : " non-feas") << "\n";);
+        if (!m_columns_with_changed_bounds.contains(j)){
```
Comment (NikolajBjorner) [inline src/math/lp/lar_solver.cpp:1734]: seems awkward to need this pre-condition

File: src/math/lp/lp_core_solver_base.h [hunk 2]
```diff
@@ -92,7 +92,7 @@ class lp_core_solver_base {
     vector<unsigned>      m_trace_of_basis_change_vector; // the even positions are entering, the odd positions are leaving
     bool                  m_tracing_basis_changes;
     // these rows are changed by adding to them a multiple of the pivot row
-    u_set*                m_touched_rows;
+    indexed_uint_set*                m_touched_rows;
```
Comment (NikolajBjorner) [inline src/math/lp/lp_core_solver_base.h:95]: = nullptr;

File: src/math/lp/nla_core.h [hunk 3]
```diff
@@ -84,7 +84,7 @@ class core {
     reslimit&                m_reslim;
     std::function<bool(lpvar)> m_relevant;
     vector<lemma> *          m_lemma_vec;
-    lp::u_set                m_to_refine;
+    indexed_uint_set                m_to_refine;
```
Comment (NikolajBjorner) [inline src/math/lp/nla_core.h:87]: spacing

File: src/math/lp/nla_core.h [hunk 4]
```diff
@@ -118,18 +118,16 @@ class core {
     void insert_to_refine(lpvar j);
     void erase_from_to_refine(lpvar j);
     
-    const lp::u_set&  active_var_set () const { return m_active_var_set;}
+    const indexed_uint_set&  active_var_set () const { return m_active_var_set;}
     bool active_var_set_contains(unsigned j) const { return m_active_var_set.contains(j); }
 
-    void insert_to_active_var_set(unsigned j) const { m_active_var_set.insert(j); }    
+    void insert_to_active_var_set(unsigned j) const { 
+        if (!m_active_var_set.contains(j))
```
Comment (NikolajBjorner) [inline src/math/lp/nla_core.h:125]: I created insert_fresh to replace the original insert. Then insert checks containment and is safe. You don't have to wrap the check to contains.

File: src/math/lp/lp_core_solver_base_def.h [hunk 5]
```diff
@@ -324,7 +324,8 @@ pivot_column_tableau(unsigned j, unsigned piv_row_index) {
             return false;
         }
         if (m_touched_rows!= nullptr)
-            m_touched_rows->insert(c.var());
+            if (!m_touched_rows->contains(c.var()))
```
Comment (NikolajBjorner) [inline src/math/lp/lp_core_solver_base_def.h:327]: you can skip the check for contains now

File: src/math/lp/lar_solver.cpp (hunk 1 above)
Comment (NikolajBjorner) [reply src/math/lp/lar_solver.cpp:1734]: you can skip this check now

File: src/math/lp/lar_solver.cpp [hunk 6]
```diff
@@ -619,8 +618,9 @@ namespace lp {
     }
 
     void lar_solver::add_touched_row(unsigned rid) {
-        if (m_settings.bound_propagation())
-            m_touched_rows.insert(rid);
+        if (m_settings.bound_propagation()) 
+            if (!m_touched_rows.contains(rid))
```
[the remark this answers is not in the document]
Comment (levnach) [reply src/math/lp/lar_solver.cpp:622]: That is how the 'insert()' of indexed_uint_set looks like. It assumes that the inserted int is new.
` void insert(unsigned x) {
        SASSERT(!contains(x));`

[the remark this answers is not in the document]
Comment (levnach) [reply src/math/lp/lar_solver.cpp:622]: nevermind:-)

File: src/math/lp/nla_grobner.cpp [hunk 7]
```diff
@@ -498,9 +498,8 @@ namespace nla {
     }
 
     void grobner::prepare_rows_and_active_vars() {
-        m_rows.clear();
-        m_rows.resize(m_lar_solver.row_count());
-        c().clear_and_resize_active_var_set();
+        m_rows.reset();
+        c().clear_active_var_set();
```
Comment (NikolajBjorner) [inline src/math/lp/nla_grobner.cpp:502]: why remove m_rows.resize?

File: src/math/lp/random_updater_def.h [hunk 8]
```diff
@@ -45,7 +44,9 @@ bool random_updater::shift_var(unsigned j) {
     if (ret) {
         const auto & A = m_lar_solver.A_r();
         for (const auto& c : A.m_columns[j]) {
-            m_var_set.erase(m_lar_solver.r_basis()[c.var()]);
+            unsigned k = m_lar_solver.r_basis()[c.var()];
+            if (m_var_set.contains(k))
```
Comment (NikolajBjorner) [inline src/math/lp/random_updater_def.h:48]: remove check to m_var_set.contains?

File: src/math/lp/nla_grobner.cpp (hunk 7 above)
Comment (levnach) [reply src/math/lp/nla_grobner.cpp:502]: m_rows is an indexed_uint_set. In the latter reset() is equivalent to clear() and there is no resize().

File: src/math/lp/random_updater_def.h (hunk 8 above)
Comment (levnach) [reply src/math/lp/random_updater_def.h:48]: Here is the remove method ` void remove(unsigned x) {
        SASSERT(contains(x));`

**Pull request merged**